configfile: "./snakefile_config"
import pandas as pd
from math import ceil

#TODO fix file paths - use file to define correct paths?

#PIPELINE PREPARATION
#filling the QSEQIDS list for the input wildcard "qseqid" in the all rule
QSEQIDS = []
with open(config['query_sequence'],'r') as qseqfile:
	for line in qseqfile.readlines():
		if ">" in line:
			#TODO adjust qseq_ids in query_sequence file in order to assign a new unique name ? - until now just refseq OR local database sequence headers are allowed
			qseq_id = line.split(" ")[0].split(">")[1].split(".")[0]
			QSEQIDS.append(qseq_id)

rule all:
	input:
		expand("{qseqid}/target_sequences.svg", qseqid=QSEQIDS),
		"reciprocal_results.html",
		 "plot_amount_hits_of_target_taxon.png",
		 "plot_evalue_distribution.png",
		 "reciprocal_results.csv",
		 "reciprocal_results_info.txt",
		 "query_sequence_information.html",
		"query_sequence_information.csv",
		expand("{qseqid}/results_rbhs.html",qseqid=QSEQIDS),
		"query_domains.tsf","query_domains.html"

#RULE 1
#Forward BLAST - inference of target sequences within the search space
rule forward_blast:
	input: fw_queries=config['query_sequence']
	output: "blastp_fw_out.table"
	params: word_size=config['fw_word_size'], e_value=config['fw_e_value'], num_alignments=config['fw_num_alignments'], num_threads=config['fw_num_threads'], database=config['blastdb']
	log:log="log/forward_blast.log"
	shell:
		"blastp -db {params.database} -outfmt \"6 qseqid sseqid pident evalue bitscore qgi sgi sacc staxids sscinames scomnames stitle slen\" -out {output} -word_size {params.word_size} -evalue {params.e_value} -num_alignments {params.num_alignments} -num_threads {params.num_threads} -query {input.fw_queries} 2> {log}"

#RULE 2
#collects the subject id's of the forward BLAST and writes those id's into a file that is used by the bw_query_preparation rule (RULE 3)
rule fw_result_processing:
	input: fw_res="blastp_fw_out.table"
	output: gi_list="bw_queries_gi_list_for_blastdbcmd.txt"
	log: log="log/fw_result_processing.log"
	run:
		with open(log[0],'w') as logfile:
			try:
				logfile.write("INFO:working with forward blast results: writing target sequences into output file\n")
				fw_results = pd.read_table(input[0], header=None)
				with open(output[0],"w+") as out:
					for gi in fw_results[7][:].unique():
						out.write(gi+"\n")
				logfile.write(('INFO:done writing {} target sequences into output file\n'.format(len(fw_results[7][:].unique()))))
				logfile.write("DONE\n")
			except Exception as e:
				logfile.write("ERROR:{}\n".format(e))
				raise Exception("ERROR in forward blast result processing: {}".format(e))


#RULE 3
#prepares the backward query fasta file for the backward blast
rule bw_query_preparation:
	input: "bw_queries_gi_list_for_blastdbcmd.txt"
	output: "bw_queries.faa"
	params: database=config["blastdb"]
	log: log="log/bw_query_preparation.log"
	shell:
		"blastdbcmd -db {params.database} -entry_batch {input} -out {output} 2> {log}"

#RULE 4
#backward blast rule. the backward blast is divided into chunks of 100 query sequences.
rule backward_blast:
	input: bw_queries="bw_queries.faa"
	output: "blastp_bw_out.table"
	params: word_size=config['bw_word_size'],e_value=config['bw_e_value'],num_alignments=config['bw_num_alignments'],max_hsps=config['bw_max_hsps'],num_threads=config['bw_num_threads'],taxid=config['bw_taxid'],database=config['backwarddb'],intermediate_filename="inter_blast.table"
	log: log="log/backward_blast.log"
	run:
		with open(log[0],'w') as logfile:
			try:
				logfile.write("INFO:starting backward blast ...\n")
				step = 100
				sequence_dict = {}
				#loading query sequences into python dictionary
				with open(input[0],'r') as infile:
					for line in infile.readlines():
						if line.startswith(">"):
							header = line.split(" ")[0].split(">")[-1]
							sequence_dict[header] = ''
						if header:
							sequence_dict[header] += line

				#preparation of transient input files for the backward blast. maximal 100 query sequences will reside in the input fasta files
				filenames = ["bw_queries_" + str(i) + ".faa" for i in range(ceil(len(sequence_dict.keys()) / float(step)))]
				logfile.write("INFO:working with: {} input fasta files\n".format(len(filenames)))
				keys = list(sequence_dict.keys())
				start = 0
				filecounter = 0
				end = step
				while start < len(keys) and filecounter <= len(filenames):
					with open(filenames[filecounter],'w') as outfile:
						for key in keys[start:end]:
							for line in sequence_dict[key]:
								outfile.write(line)
						start = end
						end += step
						filecounter += 1

				#backward blast procedure
				inter_blast_out = open("blast.table",'w')
				for index, file in enumerate(filenames):
					logfile.write("INFO:working with input sequence file {}\n".format(index))
					shell("blastp -db {params.database} -outfmt \"6 qseqid sseqid pident evalue bitscore qgi sgi sacc staxids sscinames scomnames stitle slen\" -out {params.intermediate_filename} -taxids {params.taxid} -word_size {params.word_size} -evalue {params.e_value} -num_alignments {params.num_alignments} -max_hsps {params.max_hsps} -num_threads {params.num_threads} -query {file} 2> {log}")
					with open(params.intermediate_filename,'r') as interfilename:
						for line in interfilename.readlines():
							inter_blast_out.write(line)
				inter_blast_out.close()
				logfile.write("INFO:done with blastp on all input files, removing intermediate query files ...\n")
				#removing intermediate query files
				for file in filenames:
					shell("rm {file}")
				#removing intermediate blast table (table of the last backward blast run)
				shell("rm inter_blast.table")
				logfile.write("INFO:renaming blast.table to snakemake output: blastp_bw_out.table\n")
				#changing output filename to snakemake output name
				shell("mv blast.table {output}")
				logfile.write("DONE\n")
			except Exception as e:
				logfile.write("ERROR:{}\n".format(e))
				raise Exception("[-] ERROR during backward blast with exception: {}".format(e))

#RULE 5
#extracting reciprocal best hits with pandas blast tables
rule reciprocal_best_hits:
	input: bw_res="blastp_bw_out.table", fw_res="blastp_fw_out.table"
	output: rec_best_hits="reciprocal_best_hits_protein_ids.txt"
	log: log="log/reciprocal_best_hits.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/extract_reciprocal_best_hits.py"

#RULE 6
#transforms the forward blast and backward blast tables with the rbhs table to a result output table with additional taxonomic informations.
rule blast_tables_to_csv:
	input: fw_res="blastp_fw_out.table", rec_res="reciprocal_best_hits_protein_ids.txt", query_file=config['query_sequence']
	output: result_csv="reciprocal_results.csv", taxonomy_result_csv='reciprocal_results_with_taxonomy.csv'
	log: log="log/blast_tables_to_csv.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/blast_tables_to_orthologous_table.py"

#RULE 7
#transforming csv files to html files for the frontend
rule blast_tables_to_html:
	input: rec_res="reciprocal_results_with_taxonomy.csv"
	output: rec_html="reciprocal_results.html"
	log: log="log/blast_tables_to_html.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/blast_tables_to_html.py"

#RULE 8
#outputs a summary for the RBH inference and constructs directories for each query sequence
rule build_folders_with_hit_info_for_each_qseqid:
	input: result_csv="reciprocal_results.csv", query_file=config['query_sequence']
	output: hit_information="reciprocal_results_info.txt",queries=expand("{qseqid}/target_sequence_ids.txt",qseqid=QSEQIDS),
		rec_res_html=expand("{qseqid}/results_rbhs.html",qseqid=QSEQIDS),
		result_statistics=expand("{qseqid}/basic_statistics.png",qseqid=QSEQIDS)
	params:	project_id=config['project_id'],
			result_statistics_static=expand("../../../static/images/result_images/" + str(config['project_id']) + "/{qseqid}_statistics.png",qseqid=QSEQIDS)
	log: log="log/build_folders_with_hit_info_for_each_qseqid.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/build_folders_with_hit_info_for_each_qseqid.py"

#RULE 9
#extracts sequences of RBHs from the forward database
rule extract_subject_sequences_from_database:
	input: "{qseqid}/target_sequence_ids.txt"
	output: "{qseqid}/target_sequences_raw.faa"
	params: database=config["blastdb"]
	log: log="log/{qseqid}/extract_subject_sequences_from_database.log"
	shell:
		 "blastdbcmd -db {params.database} -entry_batch {input} -out {output} 2> {log.log}"

#RULE 16
#checks if the query sequence resides in the target_sequences.faa file
rule check_if_query_sequence_is_in_target_sequences_faa:
	input: target_file=expand("{qseqid}/target_sequences_raw.faa",qseqid=QSEQIDS), query_file=config['query_sequence']
	output: expand("{qseqid}/target_sequences.faa",qseqid=QSEQIDS)
	params: qseqs=QSEQIDS
	log: log="log/check_if_query_sequence_is_in_target_sequences_faa.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/check_if_query_sequence_is_in_target_sequences_faa.py"

#RULE 10
#creates plots for the visualization of pipeline results
rule blast_tables_to_plots:
	input: fw_res="blastp_fw_out.table", rec_res="reciprocal_best_hits_protein_ids.txt",query_file=config['query_sequence']
	params:
		plot_hits_organisms_png="../../../static/images/result_images/" + str(config['project_id']) + '/plot_amount_hits_of_target_taxon.png',
		plot_evalues="../../../static/images/result_images/" + str(config['project_id']) + '/plot_evalue_distribution.png'
	output: taxids_hits_plot="plot_amount_hits_of_target_taxon.png",evalue_plot="plot_evalue_distribution.png"
	log: log="log/blast_tables_to_plots.log"
	script:
		  "../../../static/snakefiles/reciprocal_blast/plot_reciprocal_results.py"

#RULE 11
#creates a html table with information of the query sequences
rule query_sequences_to_html_table:
	input: target_file=config['query_sequence']
	params: email=config['user_email']
	output: output_html="query_sequence_information.html", output_csv="query_sequence_information.csv"
	log: log="log/query_sequences_to_html_table.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/query_sequences_to_html_table.py"

#RULE 12
#this rule shouldn't throw exceptions, if an error occurs an empty file is created
rule conduct_multiple_sequence_alignment_mafft:
	input: "{qseqid}/target_sequences.faa"
	output: "{qseqid}/target_sequences.msa"
	log: log="log/{qseqid}/msa.log"
	run:
		execute_mafft=True
		#check if target sequence file is empty (or just contains the query sequence)
		#if this is the case, there is no need for producing a msa
		with open(input[0],'r') as infile:
			if len(infile.readlines()) <= 1:
				execute_mafft=False
		try:
			if execute_mafft==True:
				#mafft execution
				mafft=shell("mafft --auto {input} > {output} 2> {log}")
				with open(log['log'],'a') as lfile:
					lfile.write("INFO:ending mafft multiple sequence alignment\n")
					if mafft != 0:
						lfile.write("ERROR:mafft returncode is != 0\n")
						raise Exception("mafft returncode is != 0\n")
					lfile.write("DONE")
			else:
				with open(log['log'],'w') as lfile:
					lfile.write("WARNING: producing empty alignment file with touch\n")
					lfile.write("DONE\n")
				#produce empty file for snakemake
				shell("touch {output}")
		except Exception as e:
			with open(log['log'],'w') as lfile:
				lfile.write("ERROR: in mafft with exception: {}\n".format(e))
				shell("touch {output}")

#RULE 13
#this rule shouldnt throw exceptions, if an error occurs an empty file is created
rule conduct_phylogeny_fast_tree:
	input: "{qseqid}/target_sequences.msa"
	output: "{qseqid}/target_sequences.tree"
	log: log="log/{qseqid}/phylogeny.log"
	run:
		#check if multiple sequence alignment has entries
		execute_fasttree=True
		with open(input[0],"r") as infile:
			if len(infile.readlines()) <= 1:
				#if there are no entries in the multiple sequence alignment there is no need for
				# conducting fasttree
				execute_fasttree == False
		try:
			#fasttree execution
			if execute_fasttree == True:
				fasttree = shell("fasttree {input} > {output} 2> {log}")
				with open(log['log'],'a') as lfile:
					lfile.write("INFO:ending fasttree phylogeny construction\n")
					if fasttree != 0:
						lfile.write("ERROR:fasttree returncode is != 0\n")
						raise Exception("fasttree returncode is != 0\n")
					lfile.write("DONE\n")
			else:
				with open(log['log'],'w') as lfile:
					lfile.write("WARNING: producing empty tree file with touch\n")
					lfile.write("DONE\n")
				#produce empty file for snakemake
				shell("touch {output}")
		except Exception as e:
			with open(log['log'],'w') as lfile:
				lfile.write("ERROR: in fasttree with exception: {}\n".format(e))
				shell("touch {output}")

#RULE 14
rule update_tree_with_taxid:
	input: nw="{qseqid}/target_sequences.tree", df="reciprocal_results_with_taxonomy.csv"
	output: "{qseqid}/transformed_treefile.nwk"
	log: log="log/{qseqid}/update_tree_with_taxid.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/update_seq_id_taxinfo.py"

#RULE 15
rule ete3_tree_to_png:
	input: tree="{qseqid}/transformed_treefile.nwk", nw="{qseqid}/target_sequences.tree"
	params: static_pic="../../../static/images/result_images/" + str(config['project_id']) + "/{qseqid}_target_sequences.svg"
	output: pic="{qseqid}/target_sequences.svg"
	log: log="log/{qseqid}/ete3_tree_to_png.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/ete3_tree_to_png.py"

#RULE 16
rule cdd_domain_rpsblast:
	input: queries=config['query_sequence']
	params: e_value=config['fw_e_value'], num_threads=config['fw_num_threads']
	output: domain_df="query_domains.tsf"
	log: log="log/cdd_domain_rpsblast.log"
	shell:
		"rpsblast -query {input.queries} -db ../databases/CDD/Cdd -outfmt \"6 qseqid qlen sacc slen qstart qend sstart send qseq sseq bitscore evalue pident stitle\" -out {output.domain_df} -evalue {params.e_value} -num_threads {params.num_threads} 2> {log.log}"

#RULE 17
rule query_domains_to_html_table:
	input: query_domains="query_domains.tsf"
	output: domain_html_table="query_domains.html"
	log: log="log/query_domains_to_html_table.log"
	script:
		"../../../static/snakefiles/reciprocal_blast/query_sequence_domains_to_html_table.py"