{% extends "base.html" %}

{% block title %} CATHI Dashboard {% endblock %}

{% block page_title %} {% endblock %}

{% block infoheader %} CATHI Dashboard {% endblock %}

{% block page_description %}
    You are logged in as: <strong>{{request.user}}</strong>.<br>
    Monitor and execute your current projects by entering the
    <a class="nav-link" href="{% url 'active_table_view' selected_table="reciprocal_blast_projects" %}">Reciprocal Projects</a> dashboard.
    Manage and view available local databases by entering the
    <a class="nav-link" href="{% url 'active_table_view' selected_table="databases" %}">Active Databases</a> dashboard.

    If you want to create a completely new project or upload custom proteomes for BLAST databases creation, take a look at the navigation bar
    on the left hand side of this dashboard.
    <br>
    <h2> Content of this Web-Tool </h2>
    <ul>
        <li> <a href="{% url 'project_creation' %}">  Reciprocal BLAST pipeline project creation </a></li>
        <li> <a href="{% url 'one_way_project_creation' %}"> One-way BLAST project creation </a></li>
        <li> BLAST Database creation</li>
        <ul>
            <li> <a href="{% url 'refseq_transactions_dashboard' %}"> based on remotely available NCBI-Databases </a></li>
            <li> <a href="{% url 'upload_genomes'  %}"> based on locally downloaded or uploaded protein fasta files </a></li>
        </ul>
        <li> <a href="{% url 'entrez_dashboard' %}"> Search target sequences with the EDirect tool </a></li>
        <li> <a href="{% url 'species_taxids' %}"> Create files with taxonomic nodes for limiting your BLAST databases by taxonomic information </a></li>
    </ul>
{% endblock %}

{% block header %}
{% endblock %}

{% block content %}
{% load static %}
{% if domain_database.domain_database_loaded == False or domain_database.domain_database_download_task_result.status != "SUCCESS" %}
    {% if domain_database.domain_database_download_task_result %}

        {% if domain_database.domain_database_download_task_result.status == "PROGRESS" %}
            <div class="dashboard_project_list_col" id="cdd_download_progress" style="display: flex;
                                                                          flex-direction: column;
                                                                          align-items: center;
                                                                          justify-content: center;">
                <button class="btn btn-dark" disabled>
                    <span class="spinner-grow spinner-grow-sm"></span>
                    <p><span class="domain_database_download_progress"></span></p>
                </button>
            </div>
        {% elif domain_database.domain_database_download_task_result.status == "FAILURE" %}
            <div class="dashboard_project_list_col" id="cdd_download_failure"  style="display: flex;
                                                                          flex-direction: column;
                                                                          align-items: center;
                                                                          justify-content: center;">
                <form class="col"
                       action="{% url "delete_domain_database" %}"
                       method="POST">
                    {% csrf_token %}
                    <input type="submit" value="Download Failed: Delete Domain Database" class="btn btn-danger">
                </form>
            </div>
        {% elif domain_database.domain_database_loaded == False and domain_database.domain_database_download_task_result.status == "SUCCESS" %}
             <div class="dashboard_project_list_col" id="project_dashboard" style="display: flex;
                                                                          flex-direction: column;
                                                                          align-items: center;
                                                                          justify-content: center;">
                {% include "blast_project/form_download_domain_database.html" %}
            </div>
        {% endif %}
    {% else %}
        <div class="dashboard_project_list_col" id="project_dashboard"  style="display: flex;
                                                                          flex-direction: column;
                                                                          align-items: center;
                                                                          justify-content: center;">
            {% include "blast_project/form_download_domain_database.html" %}
        </div>
    {% endif %}
{% endif %}

    <div class="dashboard_project_list_col" style="overflow-x: hidden">
        <h3  class="col_header" style="text-align: center"> Welcome to CATHI the Comparative Analysis Tool for Homolog Identification  </h3>
        <p style="font-size: medium; text-align: center">
            Within this dashboard you can find short help sections for each analysis step CATHI provides.
        </p>

        

        <div class="row justify-content-md-center" style="text-align: center">
            <div class="col-md">
                <button class="btn btn-dark" onclick="displayDivElement('overview')">1. Overview</button>

            </div>
            <div class="col-md">
                <button class="btn btn-dark" onclick="displayDivElement('target_search')">2. Target Sequence Search</button>
            </div>
            <div class="col-md">
                <button class="btn btn-dark" onclick="displayDivElement('database_management')">3. Database Management</button>
            </div>
            <div class="col-md">
                <button class="btn btn-dark disabled">4. Starting CATHIs Pipeline</button>
            </div>
        </div>
    </div>

    <div id="database_management" class="dashboard_project_list_col", style="overflow-x:hidden;display:none;">
        <h3 class="col_header" style="text-align:center;text-size-adjust:auto;margin-bottom:30px">Database Management</h3>
        <div class="row" style="text-align:center;justify-content:center">
            <p style="font-size:large; text-align:left">
                There are two possibilities to create BLAST databases through CATHIs dahsboards.
                Firstly, CATHI enables the download and processing of GenBank or RefSeq genome entries directly from the NCBI FTP-Servers.
                Secondly, users can upload their own FASTA-formatted proteome files, either based on concatenation of all available FASTA-, in which case
                users also have to provide metadata information, or by uploading single FASTA-files in combination with a valid scientific taxonomic name.
            </p>
            <br>


            <img src="{% static '/images/blast_database_details.jpg' %}" alt="My image">

        </div>
    </div>


    <div id="target_search" class="dashboard_project_list_col" style="overflow-x: hidden; display: none;">
        <h3 class="col_header" style="text-align: center;text-size-adjust: auto;margin-bottom: 30px"> Target Search - How can you obtain your target sequences? </h3>
        <div class="row" style="text-align:center;justify-content: center;">
            <p style="font-size: large; text-align: left">
                CATHI integrates NCBIs <a target="_blank" href="https://www.ncbi.nlm.nih.gov/books/NBK179288/"> EDirect software </a>, which can be used for searching through various NCBI databases.
                You can use this interface through CATHIs <a href="{% url 'entrez_dashboard' %}">entrez search dashboard</a>.
                Most often you will be interested in searching for specific protein sequences that you can use within CATHIs reciprocal BLAST
                or normal BLAST pipelines. For this you can do following:
            </p>
        </div>
        <br>
        <div style="font-size: large; text-align: left">
             <ol>
                 <li>Go to the <a target="_blank" href="{% url 'entrez_dashboard' %}">entrez search dashboard</a> </li>
                <li>Specify the PROTEIN database</li>
                 <li>Enter an ENTREZ search term e.g.: <b> c-di-GMP AND cyclase AND pseudomonas[ORGN] </b> </li>
                <li>Click on submit</li>
                <li>Wait until the search process is finished, this will be indicated automatically within the dashboard</li>
                <li>Check the result table and select entries by clicking on the corresponding row</li>
                <li>Click on the download button to download your selected sequences</li>
                <li>All selected sequences will get downloaded and formatted into a FASTA file</li>
                <li>Keep in mind, that you can just use target sequences derived from one particular proteome within the reciprocal BLAST pipeline</li>
            </ol>
        </div>
        <br>

        <div class="row" style="text-align:center;justify-content:center;">
            <h3 class="col_header" style="text-align: center;text-size-adjust: auto;margin-bottom: 30px"> Entrez Protein Search Results Example </h3>
            <br>
            <img src="{% static '/images/entrez_protein_db_search_results.PNG' %}" alt="Entrez Protein Search Results">
        </div>

        <br>

        <div class="row" style="text-align:center;justify-content:center;">
            <p style="font-size: large; text-align: left">
                An alternative approach is to search for paper entries within the PubMed database
                and download the associated protein sequences. This can be done through the result page of the
                PubMed entrez search result. On the top-right you can see a button: <b> Download All Associated
                    Proteins </b>, by pressing this button CATHI will link all paper entries to the protein database.
                In a subsequent step, CATHI will download the proteins as a FASTA file. This process can consume a lot of
                time, due to the linking process but can yield interesting results.
            </p>
        </div>
    </div>

    <div id="overview" class="dashboard_project_list_col" style="overflow-x: hidden;display: none;">
        <h3  class="col_header" style="text-align: center;text-size-adjust: auto;margin-bottom: 30px"> Overview Section - What can you do with CATHI?  </h3>
        <div class="row" style="text-align:center;justify-content: center;">
            <p style="font-size: large; text-align: left">
                CATHI can aid researchers in the identification of possible target sequences and to elucidate the
                evolutionary background of those sequences by generating interactive result graphs and tables similar to the example plot
                you can see in this short overview.
            </p>

            <br>
            <br>

            <p style="font-size: large; text-align: left">
                Press the button to view an example interactive visualization.
                Klick on the button to hide/show the example plot: <button class="btn btn-dark" onclick="displayDivElement('bokeh_example')">Plot Example</button>
                <br>
            </p>
        </div>

        <div class="row" style="text-align:center;justify-content: center;display:none;"  id="bokeh_example">
            {% include "blast_project/interactive_bokeh_plot_example.html" %}
            <p style="font-size: large; text-align: left">
                <br>
                You can filter the reciprocal best hits (RBHs) based on their taxonomic identity by clicking on the taxonomy
                field names of the right selection menus. In addition, you can select multiple query sequences, change the x- and y-axis
                attributes, download selected sequences or conduct a phylogenetic inference.
                <br>
            </p>
        </div>
        <p style="font-size: large; text-align: left">
            Typically, a target sequence search also involves paper search.
            With CATHI, you can utilize the E-Direct Software from NCBI, which enables searching in biological databases and in
            the pubmed database, through the Entrez Search Dashboard page.
            There, you can search for manuscripts within the PubMed database, download paper-associated protein sequences,
            or simply view each entry on NCBI.
            Furthermore, you can conveniently search and download protein sequences to serve as query sequences.
        </p>

        <br>
        <br>

        <p style="font-size: large; text-align: left">
            CATHI's main aim is to conduct a reciprocal BLAST project using user-provided target sequences.
            Therefore, users can create BLAST databases locally using the NCBI Database Dashboard or choose remotely available databases from NCBI.
            The reciprocal BLAST project is a pipeline that comprises several tasks, starting with BLAST searches and concluding with the identification of conserved protein domains.
        </p>
        <br>
        <div style="font-size: large; text-align: left">
             <ol>
                <li>Forward BLAST</li>
                <li>Backward BLAST</li>
                <li>Inference of Reciprocal Best Hits</li>
                <li>Resulting Parsing and Taxonomy Inference</li>
                <li>Generation of Interactive Tables and Graphs</li>
                <li>Multiple Sequence Alignment</li>
                <li>Trimming of Multiple Sequence Alignments</li>
                <li>Phylogenetic Inference</li>
                <li>Conserved Domain Inference</li>
            </ol>
        </div>

        <br>
        <br>
        <p style="font-size: large; text-align: left">
             With CATHI's Project Creation Dashboard users can easily create reciprocal BLAST projects.
        </p>

    </div>

{% endblock %}
{% block scripts %}
<script>
    (function () {
        var status = $('.domain_database_download_progress'),
            poll = function () {
                jQuery.ajax({
                    url: "{% url 'domain_database_download_status' %}",
                    dataType: 'json',
                    type: 'get',
                    success: function (data) { // check if available
                        if (data) { // get and check data value
                            status.text(data.progress_status); // get and print data string
                            if (data.progress_status == "SUCCESS") {
                                status.text("DONE");
                                clearInterval(pollInterval); // optional: stop poll function
                            } else if (data.progress_status == "FAILURE") {
                                status.text("ERROR");
                                clearInterval(pollInterval);
                            }

                        }
                    },
                    error: function () { // error logging
                        status.text("error polling progress data")
                    }
                });
            },
            pollInterval = setInterval(function () { // run function every 5000 ms
                poll();
            }, 5000);
        poll(); // also run function on init
    })();
</script>
{% endblock %}

