{% extends "base.html" %}

{% block title %} CATHI Project Creation {% endblock %}

{% block page_title %} Reciprocal Blast Project Creation {% endblock %}

{% block infoheader %} Project Creation {% endblock %}

{% block page_description %}
    This website can be used for creation of reciprocal BLAST sequence similarity searches. There are two
    options that can be used.
    The first option allows you to create reciprocal BLAST projects based on <strong>local protein databases</strong> and the second
    option allows you to create
    BLAST projects based on databases that reside on <strong>NCBI-Servers</strong>. The second option is achieved by
    using the <strong>-remote</strong>
    parameter of the <a href="https://www.ncbi.nlm.nih.gov/books/NBK279690/">BLAST-C++</a> software package.
    <br>
    <h2> Information on advanced BLAST settings </h2>
    <ul>
        <li> Turn the <strong>Thread</strong> parameter to the number of your available CPU cores to enhance the search
            speed (only available for local BLAST searches)
        </li>
        <li> If you search against very small local databases increase the <strong> E-Value </strong> parameter (<a
                href="https://www.metagenomics.wiki/tools/blast/evalue">E-Value informations</a>)
        </li>
        <li> Changing the <strong>Word Size</strong> parameter will effect the amount of High-Scoring-Pairs and
            increases (higher word-size) or decreases (lower word-size) the search speed
        </li>
        <li> You can provide an <strong>Entrez Query</strong> to your remote BLAST projects, this will limit your BLAST
            search based on your query (e.g. apes[organism] - more <a target="_blank"
                    href="https://www.ncbi.nlm.nih.gov/books/NBK3837/">informations</a>)
        </li>
        <li> The <strong> Max Hsps </strong> parameter is used to control the amount of HSPs on your subject sequences.
            HSPs are assigned if the scoring value between the aligned sequences is above a certain threshold, thus
            subject sequences with more than one region of similarity will have multiple HSPs.
        </li>
        <li>
            The <strong>Bitscore filter</strong> can be applied to retain high quality hits and omit putative paralogous genes.
            A bitscore of around 200 is a good value for high quality hits.
        </li>
        <li>
            The maximal amount of RBHs in MSA and Phylogeny parameter is a convenient parameter to restrict the number of
            taxonomic units that will be displayed within the phylogeny.
        </li>
        <li>
            Descriptions for the TrimAL parameters can be viewed on the corresponding tool
            <a target="_blank" href="https://trimal.readthedocs.io/en/latest/">homepage</a>.
        </li>
        <li>
            Descriptions for the Multiple Sequence Alignment visualization settings can be found
            <a target="_blank" href="https://desmid.github.io/mview/">here</a>.
        </li>
    </ul>
{% endblock %}

{% block header %}
{% endblock %}
{% block extend_body %} {% if RemoteProjectCreationForm.errors %} onload="loadRemoteOrLocal('remote')" {% endif %}
    {% if ProjectCreationForm.errors %} onload="loadRemoteOrLocal('local')" {% endif %} {% endblock %}
{% block content %}
<div id="post_form_container">
    <div class="dashboard_project_list_col">
        <h2 style="text-align: center; font-family: 'Courier New'"> <b> Reciprocal BLAST Project Creation Setup </b> </h2>
        <br>

        <div id="input_field_information" style="display: none">
            <ul class="list-group">
                <li class="list-group-item">
                    <b> Project Title </b> <br>
                    <p style="text-align: left;font-size: medium"> Just enter a UNIQUE project title.
                        This field is mandatory as it is needed for correct project assignment. </p>
                </li>
                <li class="list-group-item">
                    <b> Query Sequences </b> <br>
                    <p style="text-align: left;font-size: medium">Upload a query sequence fasta file. <br>
                        Sequences in this file MUST derive from only ONE organism. <br>
                        <br>
                        Example target file content:
                        <br>
                        <br>
                        >WP_087494701.1 lipopolysaccharide heptosyltransferase I [Curvibacter sp. AEP1-3]<br>
                        MGATHPASCRIGGGFALKAPQKILIVKLSSLGDVVHTMPAVQDLRAAFPRARIDWVVERGFAPLVARCEG<br>
                        VARVIPCDIRIWSKKPFAAATRAAWRSFKADLQQDAYDAVIDLQGLSKSALVAWLARTTPDGRRYAMANQ<br>
                        TEGSGYEAPTRWVADTAIALEAHVHAVERGRLLCAKALGYTPADALRFGLVGQSKALSAPLPKHVGHADN<br>
                        PFVPRKPLVALVHGTSRVDKEWPLEHWIALGRRLNDAGFGVSLPHAGARELATSQAIAAGLDEAWVLPAV<br>
                        ALDVLTDTLAHCAGVVGVDSGVSHIAVALDLPHVQIYNFDTAWRTGPAALDARGKPARQCSVYAQPCPEV<br>
                        ESVWSAWEALATPVLSR<br>
                        >WP_087495532.1 lipopolysaccharide heptosyltransferase II [Curvibacter sp. AEP1-3]<br>
                        MIKALVIAPQWIGDAVMTEPLLRRLAARGERITVAAVPWVAPVYRAMPQVAEVLVLPFARGGVQWAARRA<br>
                        YAAALRGQFAKAYVCPNSLKSALIPFWAGIPERIGYTGELRFGMLNQRLPNPPEGSRPPMVAFYSALSGE<br>
                        AGVDTDRPALQVAADAIAAVLAPRALHAQGFYVVAPGAEYGPAKRWPATHFAALVTRLDKPVLLLGSAKD<br>
                        DAVCNEIAEAVNAVRPGHCANLAGRTSLDEAVALIAGAHAMVSNDSGLMHVAAAFGVPQVAIFGSSSPLH<br>
                        TPPLNPAATVLWLKNDPSYQPPLDCAPCFKRDCPLGHTRCLNDITSDKVLSVL<br>
                        <br>
                        The sequence identifier (the genome from which you receive the sequences) have to reside in the backward database. <br>
                        A careful design of query sequences is recommended as too much starting sequences can significantly increase the running time. </p>
                </li>
                <li class="list-group-item">
                    <b> Protein Identifier </b>
                    <p style="text-align: left;font-size: medium">
                        If you dont want to upload a FASTA file containing your query sequences, you can list protein identifier
                        separated by commas. E.g.: WP_087495532.1, WP_087494701.1
                    </p>
                </li>
                <li class="list-group-item">
                    <b> Forward BLAST database </b> <br>
                    <p style="text-align: left;font-size: medium">
                        This database is the "search space" for your provided query sequences. <br>
                        This database is used for the initial BLAST, the forward BLAST. <br>
                        All hits are extracted and assembled in a new fasta file, named backward_query_sequences.fasta.
                        Those hits are used as input for the second BLAST, the backward BLAST.
                    </p>
                </li>
                <li class="list-group-item">
                    <b> Backward BLAST database </b> <br>
                    <p style="text-align: left;font-size: medium"> The organism or/and genome of the query sequences has to reside in this database.
                        All hits of the forward BLAST are used as input for the backward BLAST, which uses this database.
                        If those hits match the exact query sequence of the forward BLAST, reciprocal best hits (RBHs) are found.
                        E.g., if you want to search RBHs for WP_087494701.1, WP_087494701.1 needs to reside in your backward BLAST database (you need
                        to select a database containing the proteome of Curvibacter sp. AEP1-3). <br>
                        This field gets validated by CATHI. If you receive an error message within this field, you may need to change the selected backward BLAST database
                        to a database truly containing your query sequences. The best option is to search for proteins within one proteome, upload this proteome as
                        a local BLAST database and use this database as the backward BLAST database.
                    </p>
                </li>
                <li class="list-group-item">
                    <b> Species name for the backward BLAST </b> <br>
                    <p style="text-align: left;font-size: medium"> This is the scientific name of the organisms from which the query sequences were obtained.
                        Check the backward BLAST database if you don't know the exact name of the organism from which you want to infer RBHs.
                        E.g., if you want to search RBHs for WP_087494701.1, which is a protein sequence of Curvibacter sp. AEP1-3, you need to enter Curvibacter sp. AEP1-3.
                    </p>
                </li>
            </ul>
            <br>
        </div>
        <button class="btn btn-dark" onclick="displayDivElement('input_field_information')" style="position: absolute;top: 10px; right: 10px;"> Input Field Info </button>
    </div>

    {% if taxdb == True %}
        {% if domain_database.domain_database_loaded == False or domain_database.domain_database_download_task_result.status != "SUCCESS" %}
            <div class="dashboard_project_list_col" >
                <p style="text-align: center"> You have to download the Conserved Domain Database (CDD) before you can create any reciprocal BLAST projects.
                    Return to the Dashboard and download the CDD.</p>
            </div>
        {% else %}

            <div class="text-center dashboard_project_list_col">
                <form name="project_creation_menu" action="javascript:void(0)" id="one_way_project_creation_form">
                    <div class="custom-control custom-radio custom-control-inline">
                        <input class="custom-control-input" type="radio" id="local" name="local_search_button" value="local"
                               onchange="loadRemoteOrLocal(this.value)">
                        <label class="custom-control-label" for="local">Search Against Available Local BLAST Databases</label>
                    </div>

                    <div class="custom-control custom-radio custom-control-inline">
                        <input class="custom-control-input" type="radio" id="remote" name="remote_search_button" value="remote"
                               onchange="loadRemoteOrLocal(this.value)">
                        <label class="custom-control-label" for="remote">Search Against Available Remote BLAST Databases</label>
                    </div>
                </form>
            </div>

            <div class="dashboard_project_list_col" id="project_creation_local" style="display: none">
                {% include "blast_project/form_project_creation.html" %}
                <br>
                <button class="btn btn-dark" onclick="displayDivElement('advanced_settings')" >Advanced BLAST Settings </button>
            </div>

            <div class="dashboard_project_list_col" id="project_creation_remote" style="display: none">
                {% include "blast_project/form_remote_project_creation.html" %}
                <br>
                <button class="btn btn-dark" onclick="displayDivElement('advanced_remote_settings')">Advanced BLAST Settings</button>
            </div>

        {% endif %}
    {% endif %}
    {% if taxdb == False %}
        <div class="dashboard_project_list_col">
            <p> There is no active taxonomic database, before you can use CATHIs pipeline,
                you should download NCBIs taxonomy database and extract it into the database directory of CATHI.</p>
        </div>
    {% endif %}
</div>
{% endblock %}